"""
–†–µ–∞–ª–∏–∑–∞—Ü–∏—è TranscriptionService —Å –∏—Å–ø–æ–ª—å–∑–æ–≤–∞–Ω–∏–µ–º Borealis –º–æ–¥–µ–ª–∏.
–û–±—Ä–∞–±–∞—Ç—ã–≤–∞–µ—Ç –∞—É–¥–∏–æ —Ñ–∞–π–ª—ã –∏ –≤–æ–∑–≤—Ä–∞—â–∞–µ—Ç —Ç—Ä–∞–Ω—Å–∫—Ä–∏–ø—Ü–∏—é —Ç–µ–∫—Å—Ç–∞.

Production v4.0 (FULLY OPTIMIZED)
‚úÖ Batch Size 32 (–æ–ø—Ç–∏–º–∞–ª—å–Ω—ã–π –¥–ª—è RTX 5080)
‚úÖ Pinned Memory –¥–ª—è –±—ã—Å—Ç—Ä–æ–≥–æ –∫–æ–ø–∏—Ä–æ–≤–∞–Ω–∏—è CPU‚ÜíGPU
‚úÖ –ê—Å–∏–Ω—Ö—Ä–æ–Ω–Ω–∞—è –æ–±—Ä–∞–±–æ—Ç–∫–∞ GPU/CPU
‚úÖ CUDA Streams –∏ non-blocking transfer
‚úÖ –£–º–Ω–æ–µ —Ä–∞–∑—Ä–µ–∑–∞–Ω–∏–µ –ø–æ –ø–∞—É–∑–∞–º (–∫–æ–Ω—Ç–µ–∫—Å—Ç —Å–æ—Ö—Ä–∞–Ω–µ–Ω)
"""

import os
import sys
import time
import tempfile
import threading
from pathlib import Path
from queue import Queue

# –î–æ–±–∞–≤–ª—è–µ–º –∫–æ—Ä–Ω–µ–≤—É—é –¥–∏—Ä–µ–∫—Ç–æ—Ä–∏—é –≤ –ø—É—Ç—å –¥–ª—è –∏–º–ø–æ—Ä—Ç–æ–≤
sys.path.insert(0, str(Path(__file__).parent.parent.parent.parent))

# –û—Ç–∫–ª—é—á–∞–µ–º –æ–Ω–ª–∞–π–Ω-–∑–∞–≥—Ä—É–∑–∫—É –º–æ–¥–µ–ª–µ–π
os.environ['HF_HUB_OFFLINE'] = '1'

from generated.v1 import transcription_pb2
from generated.v1 import transcription_pb2_grpc
from services.transcription.base_service import TranscriptionServiceBase
from resources.config import config

import torch
import librosa
import numpy as np
from transformers import AutoModelForCausalLM, AutoTokenizer, AutoFeatureExtractor


class BorealisTranscriptionService(TranscriptionServiceBase, transcription_pb2_grpc.TranscriptionServiceServicer):
    """
    –†–µ–∞–ª–∏–∑–∞—Ü–∏—è TranscriptionService —Å –∏—Å–ø–æ–ª—å–∑–æ–≤–∞–Ω–∏–µ–º Borealis ML –º–æ–¥–µ–ª–∏.
    –ü—Ä–∏–Ω–∏–º–∞–µ—Ç –∞—É–¥–∏–æ —Ñ–∞–π–ª—ã –∏ –≤–æ–∑–≤—Ä–∞—â–∞–µ—Ç —Ç—Ä–∞–Ω—Å–∫—Ä–∏–ø—Ü–∏—é —Å –∏—Å–ø–æ–ª—å–∑–æ–≤–∞–Ω–∏–µ–º –º–æ–¥–µ–ª–∏ Borealis (Vikhrmodels/Borealis).

    –û–ø—Ç–∏–º–∏–∑–∞—Ü–∏–∏ v4.0:
    - Batch Size 32 –¥–ª—è –º–∞–∫—Å–∏–º–∞–ª—å–Ω–æ–π –ø—Ä–æ–∏–∑–≤–æ–¥–∏—Ç–µ–ª—å–Ω–æ—Å—Ç–∏
    - Pinned Memory –¥–ª—è –±—ã—Å—Ç—Ä–æ–≥–æ DMA –∫–æ–ø–∏—Ä–æ–≤–∞–Ω–∏—è
    - –ê—Å–∏–Ω—Ö—Ä–æ–Ω–Ω–∞—è –æ–±—Ä–∞–±–æ—Ç–∫–∞ GPU/CPU
    - CUDA Streams –¥–ª—è –ø–∞—Ä–∞–ª–ª–µ–ª—å–Ω–æ–π —Ä–∞–±–æ—Ç—ã
    - –£–º–Ω–æ–µ —Ä–∞–∑—Ä–µ–∑–∞–Ω–∏–µ –∞—É–¥–∏–æ –ø–æ –ø–∞—É–∑–∞–º
    """

    def __init__(self):
        """–ò–Ω–∏—Ü–∏–∞–ª–∏–∑–∞—Ü–∏—è Borealis —Å–µ—Ä–≤–∏—Å–∞ —Ç—Ä–∞–Ω—Å–∫—Ä–∏–ø—Ü–∏–∏"""
        super().__init__()

        self.logger.info("=" * 80)
        self.logger.info("BorealisTranscriptionService - Production v4.0 (FULLY OPTIMIZED)")
        self.logger.info("=" * 80)

        # –ó–∞–≥—Ä—É–∑–∫–∞ –∫–æ–Ω—Ñ–∏–≥—É—Ä–∞—Ü–∏–∏
        self.logger.info("–ó–∞–≥—Ä—É–∑–∫–∞ –∫–æ–Ω—Ñ–∏–≥—É—Ä–∞—Ü–∏–∏...")
        self.logger.info(f"  MODEL_NAME: {config.MODEL_NAME}")
        self.logger.info(f"  DEVICE: {config.DEVICE}")
        self.logger.info(f"  BATCH_SIZE: {config.BATCH_SIZE}")
        self.logger.info(f"  TARGET_CHUNK_DURATION: {config.TARGET_CHUNK_DURATION}s")
        self.logger.info(f"  LOCAL_FILES_ONLY: {config.MODEL_LOCAL_FILES_ONLY}")

        # –ó–∞–≥—Ä—É–∑–∫–∞ –º–æ–¥–µ–ª–∏ Borealis
        self.logger.info("–ó–∞–≥—Ä—É–∑–∫–∞ –º–æ–¥–µ–ª–∏ Borealis...")

        self.model = AutoModelForCausalLM.from_pretrained(
            config.MODEL_NAME,
            trust_remote_code=True,
            local_files_only=config.MODEL_LOCAL_FILES_ONLY
        )
        self.tokenizer = AutoTokenizer.from_pretrained(config.MODEL_NAME, local_files_only=config.MODEL_LOCAL_FILES_ONLY)
        self.extractor = AutoFeatureExtractor.from_pretrained(config.MODEL_NAME, local_files_only=config.MODEL_LOCAL_FILES_ONLY)

        self.model.eval()
        self.model.to(config.DEVICE)
        self.model = torch.compile(self.model, mode="reduce-overhead", fullgraph=False)

        self.logger.info(f"‚úì –ú–æ–¥–µ–ª—å –∑–∞–≥—Ä—É–∂–µ–Ω–∞ –Ω–∞ {next(self.model.parameters()).device}")

        # CUDA streams
        self.compute_stream = torch.cuda.default_stream()
        self.transfer_stream = torch.cuda.Stream()

        # –ü–∞—Ä–∞–º–µ—Ç—Ä—ã –≥–µ–Ω–µ—Ä–∞—Ü–∏–∏
        self.generation_params = {
            "max_new_tokens": 350,
            "do_sample": True,
            "top_p": 0.9,
            "top_k": 50,
            "temperature": 0.2,
            "use_cache": True,
        }

        self.logger.info("‚úì BorealisTranscriptionService –∏–Ω–∏—Ü–∏–∞–ª–∏–∑–∏—Ä–æ–≤–∞–Ω")
        self.logger.info("=" * 80)

    def _find_optimal_cut_points(self, waveform, sr, target_chunk_duration=None, window_duration=5):
        """–ù–∞—Ö–æ–¥–∏—Ç –æ–ø—Ç–∏–º–∞–ª—å–Ω—ã–µ —Ç–æ—á–∫–∏ —Ä–∞–∑—Ä–µ–∑–∞–Ω–∏—è –ø–æ —ç–Ω–µ—Ä–≥–∏–∏"""
        if target_chunk_duration is None:
            target_chunk_duration = config.TARGET_CHUNK_DURATION

        self.logger.info("–ê–Ω–∞–ª–∏–∑: –ø–æ–∏—Å–∫ –æ–ø—Ç–∏–º–∞–ª—å–Ω—ã—Ö —Ç–æ—á–µ–∫ —Ä–∞–∑—Ä–µ–∑–∞–Ω–∏—è...")

        frame_length = 2048
        hop_length = 512

        S = librosa.stft(waveform, n_fft=frame_length, hop_length=hop_length)
        magnitude = np.abs(S)
        energy = np.sum(magnitude ** 2, axis=0)
        energy = (energy - np.min(energy)) / (np.max(energy) - np.min(energy) + 1e-10)

        def frame_to_time(frame_idx):
            return (frame_idx * hop_length) / sr

        def time_to_frame(time_s):
            return int((time_s * sr) / hop_length)

        total_duration = len(waveform) / sr
        cut_points = []

        for i in range(1, int(total_duration / target_chunk_duration) + 1):
            target_time = i * target_chunk_duration
            if target_time >= total_duration:
                break

            window_frames = int((window_duration * sr) / hop_length)
            target_frame = time_to_frame(target_time)
            start_frame = max(0, target_frame - window_frames // 2)
            end_frame = min(len(energy), target_frame + window_frames // 2)

            window_energy = energy[start_frame:end_frame]
            min_energy_idx = np.argmin(window_energy)
            best_frame = start_frame + min_energy_idx
            best_time = frame_to_time(best_frame)

            cut_points.append(best_time)

        self.logger.info(f"‚úì –ù–∞–π–¥–µ–Ω–æ {len(cut_points)} —Ç–æ—á–µ–∫ —Ä–∞–∑—Ä–µ–∑–∞–Ω–∏—è")
        return cut_points

    def _split_audio_by_cut_points(self, waveform, sr, cut_points):
        """–†–∞–∑–±–∏–≤–∞–µ—Ç –∞—É–¥–∏–æ –ø–æ —Ç–æ—á–∫–∞–º"""
        chunks = []
        prev_pos = 0

        for best_time in cut_points:
            cut_sample = int(best_time * sr)
            chunk = waveform[prev_pos:cut_sample]
            if len(chunk) > 0:
                chunks.append(chunk)
            prev_pos = cut_sample

        if prev_pos < len(waveform):
            chunks.append(waveform[prev_pos:])

        return chunks

    def _prepare_batch_pinned(self, chunks, start_idx, batch_size, sr):
        """
        –ü–æ–¥–≥–æ—Ç–∞–≤–ª–∏–≤–∞–µ—Ç –±–∞—Ç—á –≤ Pinned Memory
        –≠—Ç–æ –ø–æ–∑–≤–æ–ª—è–µ—Ç DMA –∫–æ–ø–∏—Ä–æ–≤–∞–Ω–∏–µ (–±—ã—Å—Ç—Ä–µ–µ —Å CPU –Ω–∞ GPU)
        """
        batch = chunks[start_idx:start_idx+batch_size]

        mel_batch = []
        att_mask_batch = []

        for chunk in batch:
            proc = self.extractor(chunk, sampling_rate=sr, padding="max_length",
                                 max_length=480_000, return_attention_mask=True, return_tensors="pt")
            mel_batch.append(proc.input_features.squeeze(0))
            att_mask_batch.append(proc.attention_mask.squeeze(0))

        mel = torch.stack(mel_batch)
        att_mask = torch.stack(att_mask_batch)

        # ‚úÖ –ó–∞—Ñ–∏–∫—Å–∏—Ä—É–µ–º –≤ Pinned Memory
        mel = mel.pin_memory()
        att_mask = att_mask.pin_memory()

        return mel, att_mask

    def _process_chunks_v4(self, chunks, sr, batch_size=None):
        """
        –û–ø—Ç–∏–º–∏–∑–∏—Ä–æ–≤–∞–Ω–Ω–∞—è –∞—Å–∏–Ω—Ö—Ä–æ–Ω–Ω–∞—è –æ–±—Ä–∞–±–æ—Ç–∫–∞ v4.0

        ‚úÖ Batch Size –∏–∑ –∫–æ–Ω—Ñ–∏–≥—É—Ä–∞—Ü–∏–∏
        ‚úÖ Pinned Memory
        ‚úÖ –ê—Å–∏–Ω—Ö—Ä–æ–Ω–Ω–æ–µ –∫–æ–ø–∏—Ä–æ–≤–∞–Ω–∏–µ
        ‚úÖ GPU + CPU –ø–∞—Ä–∞–ª–ª–µ–ª—å–Ω–æ
        """
        if batch_size is None:
            batch_size = config.BATCH_SIZE

        results = []
        results_lock = threading.Lock()
        batch_queue = Queue(maxsize=2)
        stop_event = threading.Event()

        def gpu_worker():
            """–ü–æ—Ç–æ–∫ –æ–±—Ä–∞–±–æ—Ç–∫–∏ GPU"""
            try:
                while not stop_event.is_set():
                    try:
                        item = batch_queue.get(timeout=1)
                        if item is None:
                            break

                        mel, att_mask, batch_indices = item

                        # –ê—Å–∏–Ω—Ö—Ä–æ–Ω–Ω–æ–µ –∫–æ–ø–∏—Ä–æ–≤–∞–Ω–∏–µ –≤ –æ—Ç–¥–µ–ª—å–Ω–æ–º stream
                        with torch.cuda.stream(self.transfer_stream):
                            mel = mel.to("cuda", non_blocking=True)
                            att_mask = att_mask.to("cuda", non_blocking=True)

                        # –°–∏–Ω—Ö—Ä–æ–Ω–∏–∑–∏—Ä—É–µ–º
                        self.transfer_stream.synchronize()

                        # –û–±—Ä–∞–±–æ—Ç–∫–∞ –Ω–∞ GPU
                        with torch.inference_mode():
                            transcripts = self.model.generate(
                                mel=mel, att_mask=att_mask,
                                **self.generation_params
                            )

                        with results_lock:
                            for idx, transcript in zip(batch_indices, transcripts):
                                results.append((idx, str(transcript)))

                    except Exception as e:
                        self.logger.error(f"GPU worker –æ—à–∏–±–∫–∞: {e}")
            except Exception as e:
                self.logger.error(f"GPU worker fatal: {e}")

        # –ó–∞–ø—É—Å–∫–∞–µ–º GPU –ø–æ—Ç–æ–∫
        gpu_thread = threading.Thread(target=gpu_worker, daemon=False)
        gpu_thread.start()

        # ========== MAIN LOOP (CPU) ==========
        self.logger.info(f"–û–±—Ä–∞–±–æ—Ç–∫–∞: v4.0 - –ê—Å–∏–Ω—Ö—Ä–æ–Ω–Ω–∞—è –æ–±—Ä–∞–±–æ—Ç–∫–∞ {len(chunks)} –∫—É—Å–∫–æ–≤")
        self.logger.info(f"           Batch Size: {batch_size} | Pinned Memory ‚úì | GPU + CPU –ø–∞—Ä–∞–ª–ª–µ–ª—å–Ω–æ ‚ö°")

        for batch_start_idx in range(0, len(chunks), batch_size):
            # CPU –ø–æ–¥–≥–æ—Ç–∞–≤–ª–∏–≤–∞–µ—Ç –±–∞—Ç—á –ø–æ–∫–∞ GPU —Ä–∞–±–æ—Ç–∞–µ—Ç
            mel, att_mask = self._prepare_batch_pinned(chunks, batch_start_idx, batch_size, sr)
            batch_indices = list(range(batch_start_idx, min(batch_start_idx + batch_size, len(chunks))))

            # –û—Ç–ø—Ä–∞–≤–ª—è–µ–º –≤ –æ—á–µ—Ä–µ–¥—å
            batch_queue.put((mel, att_mask, batch_indices))

            # –ü—Ä–æ–≥—Ä–µ—Å—Å
            processed = min(batch_start_idx + batch_size, len(chunks))
            progress = (processed / len(chunks)) * 100
            self.logger.info(f"  –ü—Ä–æ–≥—Ä–µ—Å—Å: {processed}/{len(chunks)} ({progress:.0f}%)")

        # –ó–∞–≤–µ—Ä—à–µ–Ω–∏–µ
        batch_queue.put(None)
        gpu_thread.join()

        # –°–æ—Ä—Ç–∏—Ä—É–µ–º —Ä–µ–∑—É–ª—å—Ç–∞—Ç—ã –≤ –ø—Ä–∞–≤–∏–ª—å–Ω–æ–º –ø–æ—Ä—è–¥–∫–µ
        results.sort(key=lambda x: x[0])
        return [text for _, text in results]

    def _transcribe_audio_file(self, audio_path):
        """
        –ü–æ–ª–Ω–∞—è –æ–±—Ä–∞–±–æ—Ç–∫–∞ –∞—É–¥–∏–æ —Ñ–∞–π–ª–∞ —Å –∏—Å–ø–æ–ª—å–∑–æ–≤–∞–Ω–∏–µ–º Borealis –º–æ–¥–µ–ª–∏

        Args:
            audio_path: –ü—É—Ç—å –∫ –∞—É–¥–∏–æ —Ñ–∞–π–ª—É

        Returns:
            –¢—Ä–∞–Ω—Å–∫—Ä–∏–ø—Ü–∏—è —Ç–µ–∫—Å—Ç–∞
        """
        transcription_start = time.time()

        # –ó–∞–≥—Ä—É–∑–∫–∞ –∞—É–¥–∏–æ
        self.logger.info(f"–ó–∞–≥—Ä—É–∑–∫–∞ –∞—É–¥–∏–æ: {audio_path}")
        load_start = time.time()

        waveform, sr = librosa.load(audio_path, sr=16_000)
        total_duration = len(waveform) / sr
        load_time = time.time() - load_start

        self.logger.info(f"‚úì –ó–∞–≥—Ä—É–∂–µ–Ω–æ {total_duration:.1f}s –∑–∞ {load_time:.2f}s")

        # –ê–Ω–∞–ª–∏–∑ –∏ —Ä–∞–∑—Ä–µ–∑–∞–Ω–∏–µ
        analysis_start = time.time()
        cut_points = self._find_optimal_cut_points(waveform, sr)
        chunks = self._split_audio_by_cut_points(waveform, sr, cut_points)
        analysis_time = time.time() - analysis_start

        self.logger.info(f"‚úì –†–∞–∑–±–∏—Ç–æ –Ω–∞ {len(chunks)} –∫—É—Å–∫–æ–≤ –∑–∞ {analysis_time:.2f}s")

        # –û–±—Ä–∞–±–æ—Ç–∫–∞
        process_start = time.time()
        results = self._process_chunks_v4(chunks, sr)
        process_time = time.time() - process_start

        full_transcript = " ".join(results)

        self.logger.info(f"‚úì –û–±—Ä–∞–±–æ—Ç–∞–Ω–æ –∑–∞ {process_time:.2f}s")
        self.logger.info(f"  –°–∫–æ—Ä–æ—Å—Ç—å: {total_duration/process_time:.1f}x –æ—Ç —Ä–µ–∞–ª—å–Ω–æ–≥–æ –≤—Ä–µ–º–µ–Ω–∏")

        # –°—Ç–∞—Ç–∏—Å—Ç–∏–∫–∞
        total_time = time.time() - transcription_start
        words = len(full_transcript.split())
        chars = len(full_transcript)
        gpu_mem = torch.cuda.memory_allocated() / 1e9

        self.logger.info("=" * 80)
        self.logger.info("–†–ï–ó–£–õ–¨–¢–ê–¢ v4.0 FULLY OPTIMIZED")
        self.logger.info("=" * 80)
        self.logger.info(f"‚è±Ô∏è  –í—Ä–µ–º—è: –ó–∞–≥—Ä—É–∑–∫–∞={load_time:.2f}s | –ê–Ω–∞–ª–∏–∑={analysis_time:.2f}s | –¢—Ä–∞–Ω—Å–∫—Ä–∏–ø—Ü–∏—è={process_time:.2f}s | –ò–¢–û–ì–û={total_time:.2f}s")
        self.logger.info(f"üìä –¢–µ–∫—Å—Ç: –°–∏–º–≤–æ–ª–æ–≤={chars} | –°–ª–æ–≤={words} | –°–∫–æ—Ä–æ—Å—Ç—å={total_duration/total_time:.1f}x")
        self.logger.info(f"üíæ GPU: –ò—Å–ø–æ–ª—å–∑–æ–≤–∞–Ω–æ={gpu_mem:.2f}GB")
        self.logger.info("=" * 80)

        return full_transcript

    def TranscribeAudio(self, request, context):
        """
        –ü—Ä–∏–Ω–∏–º–∞–µ—Ç –∞—É–¥–∏–æ —Ñ–∞–π–ª –∏ –≤–æ–∑–≤—Ä–∞—â–∞–µ—Ç —Ç—Ä–∞–Ω—Å–∫—Ä–∏–ø—Ü–∏—é.

        Args:
            request: AudioRequest —Å –¥–∞–Ω–Ω—ã–º–∏ –∞—É–¥–∏–æ —Ñ–∞–π–ª–∞
            context: gRPC context

        Returns:
            TranscriptionResponse —Å —Ä–µ–∑—É–ª—å—Ç–∞—Ç–æ–º —Ç—Ä–∞–Ω—Å–∫—Ä–∏–ø—Ü–∏–∏
        """
        start_time = time.time()

        self.logger.info("=" * 80)
        self.logger.info(f"üì• –ü–æ–ª—É—á–µ–Ω –∑–∞–ø—Ä–æ—Å –Ω–∞ —Ç—Ä–∞–Ω—Å–∫—Ä–∏–ø—Ü–∏—é —Ñ–∞–π–ª–∞: {request.filename}")
        self.logger.info(f"   –†–∞–∑–º–µ—Ä —Ñ–∞–π–ª–∞: {len(request.audio_data) / (1024*1024):.2f} –ú–ë")
        self.logger.info(f"   –§–æ—Ä–º–∞—Ç: {request.format}")
        self.logger.info("=" * 80)

        # –í–∞–ª–∏–¥–∞—Ü–∏—è –∑–∞–ø—Ä–æ—Å–∞
        is_valid, error_msg = self._validate_transcription_request(request.filename, request.audio_data)

        if not is_valid:
            self.logger.error(f"‚ùå –û—à–∏–±–∫–∞ –≤–∞–ª–∏–¥–∞—Ü–∏–∏: {error_msg}")
            processing_time = time.time() - start_time

            return transcription_pb2.TranscriptionResponse(
                transcript="",
                success=False,
                error_message=error_msg,
                processing_time=processing_time,
                audio_duration=0.0,
                stats=transcription_pb2.TranscriptionStats(
                    word_count=0,
                    char_count=0,
                    speed_factor=0.0
                )
            )

        # –¢—Ä–∞–Ω—Å–∫—Ä–∏–ø—Ü–∏—è —Å –∏—Å–ø–æ–ª—å–∑–æ–≤–∞–Ω–∏–µ–º Borealis –º–æ–¥–µ–ª–∏
        try:
            # –°–æ—Ö—Ä–∞–Ω—è–µ–º –∞—É–¥–∏–æ –≤–æ –≤—Ä–µ–º–µ–Ω–Ω—ã–π —Ñ–∞–π–ª
            with tempfile.NamedTemporaryFile(suffix=f".{request.format}", delete=False) as temp_file:
                temp_file.write(request.audio_data)
                temp_audio_path = temp_file.name

            self.logger.info(f"üíæ –í—Ä–µ–º–µ–Ω–Ω—ã–π —Ñ–∞–π–ª —Å–æ–∑–¥–∞–Ω: {temp_audio_path}")

            # –ó–∞–ø—É—Å–∫–∞–µ–º —Ç—Ä–∞–Ω—Å–∫—Ä–∏–ø—Ü–∏—é —Å –∏—Å–ø–æ–ª—å–∑–æ–≤–∞–Ω–∏–µ–º –∏–Ω—Ç–µ–≥—Ä–∏—Ä–æ–≤–∞–Ω–Ω–æ–π –ª–æ–≥–∏–∫–∏
            transcript = self._transcribe_audio_file(temp_audio_path)

            # –í—ã—á–∏—Å–ª—è–µ–º –¥–ª–∏—Ç–µ–ª—å–Ω–æ—Å—Ç—å –¥–ª—è —Å—Ç–∞—Ç–∏—Å—Ç–∏–∫–∏
            waveform, sr = librosa.load(temp_audio_path, sr=16_000)
            audio_duration = len(waveform) / sr

            # –£–¥–∞–ª—è–µ–º –≤—Ä–µ–º–µ–Ω–Ω—ã–π —Ñ–∞–π–ª
            try:
                Path(temp_audio_path).unlink()
                self.logger.info(f"üóëÔ∏è  –í—Ä–µ–º–µ–Ω–Ω—ã–π —Ñ–∞–π–ª —É–¥–∞–ª–µ–Ω")
            except Exception as e:
                self.logger.warning(f"‚ö†Ô∏è  –ù–µ —É–¥–∞–ª–æ—Å—å —É–¥–∞–ª–∏—Ç—å –≤—Ä–µ–º–µ–Ω–Ω—ã–π —Ñ–∞–π–ª: {e}")

            if transcript is None or transcript == "":
                raise Exception("–¢—Ä–∞–Ω—Å–∫—Ä–∏–ø—Ü–∏—è –≤–µ—Ä–Ω—É–ª–∞ –ø—É—Å—Ç–æ–π —Ä–µ–∑—É–ª—å—Ç–∞—Ç")

            processing_time = time.time() - start_time
            word_count = len(transcript.split())
            char_count = len(transcript)
            speed_factor = audio_duration / processing_time if processing_time > 0 else 0.0

            self.logger.info("=" * 80)
            self.logger.info(f"‚úÖ –¢—Ä–∞–Ω—Å–∫—Ä–∏–ø—Ü–∏—è —É—Å–ø–µ—à–Ω–æ –∑–∞–≤–µ—Ä—à–µ–Ω–∞!")
            self.logger.info(f"   –û–±—â–µ–µ –≤—Ä–µ–º—è –æ–±—Ä–∞–±–æ—Ç–∫–∏: {processing_time:.2f} —Å–µ–∫")
            self.logger.info(f"   –°–∫–æ—Ä–æ—Å—Ç—å: {speed_factor:.1f}x –æ—Ç —Ä–µ–∞–ª—å–Ω–æ–≥–æ –≤—Ä–µ–º–µ–Ω–∏")
            self.logger.info(f"   –°–ª–æ–≤: {word_count}, –°–∏–º–≤–æ–ª–æ–≤: {char_count}")
            self.logger.info("=" * 80)

            return transcription_pb2.TranscriptionResponse(
                transcript=transcript,
                success=True,
                error_message="",
                processing_time=processing_time,
                audio_duration=audio_duration,
                stats=transcription_pb2.TranscriptionStats(
                    word_count=word_count,
                    char_count=char_count,
                    speed_factor=speed_factor
                )
            )

        except Exception as e:
            error_msg = f"–û—à–∏–±–∫–∞ –ø—Ä–∏ —Ç—Ä–∞–Ω—Å–∫—Ä–∏–ø—Ü–∏–∏: {str(e)}"
            self.logger.error(f"‚ùå {error_msg}")
            processing_time = time.time() - start_time

            # –ü—ã—Ç–∞–µ–º—Å—è —É–¥–∞–ª–∏—Ç—å –≤—Ä–µ–º–µ–Ω–Ω—ã–π —Ñ–∞–π–ª –≤ —Å–ª—É—á–∞–µ –æ—à–∏–±–∫–∏
            try:
                if 'temp_audio_path' in locals():
                    Path(temp_audio_path).unlink()
            except:
                pass

            return transcription_pb2.TranscriptionResponse(
                transcript="",
                success=False,
                error_message=error_msg,
                processing_time=processing_time,
                audio_duration=0.0,
                stats=transcription_pb2.TranscriptionStats(
                    word_count=0,
                    char_count=0,
                    speed_factor=0.0
                )
            )

    def TranscribeAudioStream(self, request_iterator, context):
        """
        –ü—Ä–∏–Ω–∏–º–∞–µ—Ç –∞—É–¥–∏–æ —Ñ–∞–π–ª —á–µ—Ä–µ–∑ —Å—Ç—Ä–∏–º –∏ –≤–æ–∑–≤—Ä–∞—â–∞–µ—Ç —Ç—Ä–∞–Ω—Å–∫—Ä–∏–ø—Ü–∏—é.

        Args:
            request_iterator: –ò—Ç–µ—Ä–∞—Ç–æ—Ä AudioChunk
            context: gRPC context

        Returns:
            TranscriptionResponse —Å —Ä–µ–∑—É–ª—å—Ç–∞—Ç–æ–º —Ç—Ä–∞–Ω—Å–∫—Ä–∏–ø—Ü–∏–∏
        """
        start_time = time.time()

        self.logger.info("=" * 80)
        self.logger.info("üì• –ü–æ–ª—É—á–µ–Ω —Å—Ç—Ä–∏–º–∏–Ω–≥–æ–≤—ã–π –∑–∞–ø—Ä–æ—Å –Ω–∞ —Ç—Ä–∞–Ω—Å–∫—Ä–∏–ø—Ü–∏—é")
        self.logger.info("=" * 80)

        filename = ""
        format_type = ""
        sample_rate = 0
        chunks_data = []
        chunk_count = 0

        try:
            for chunk in request_iterator:
                chunk_count += 1

                # –ü–µ—Ä–≤—ã–π —á–∞–Ω–∫ —Å–æ–¥–µ—Ä–∂–∏—Ç –º–µ—Ç–∞–¥–∞–Ω–Ω—ã–µ
                if chunk_count == 1:
                    filename = chunk.filename
                    format_type = chunk.format
                    sample_rate = chunk.sample_rate
                    self.logger.info(f"üìÇ –ù–∞—á–∞–ª–æ –ø—Ä–∏–µ–º–∞ —Ñ–∞–π–ª–∞: {filename}")

                # –°–æ–±–∏—Ä–∞–µ–º –¥–∞–Ω–Ω—ã–µ
                chunks_data.append(chunk.chunk_data)

            # –û–±—ä–µ–¥–∏–Ω—è–µ–º –≤—Å–µ —á–∞–Ω–∫–∏
            audio_data = b''.join(chunks_data)
            actual_size = len(audio_data)

            self.logger.info(f"‚úì –ü–æ–ª—É—á–µ–Ω–æ {chunk_count} —á–∞–Ω–∫–æ–≤, –≤—Å–µ–≥–æ {actual_size / (1024*1024):.2f} –ú–ë")

            # –í–∞–ª–∏–¥–∞—Ü–∏—è
            is_valid, error_msg = self._validate_transcription_request(filename, audio_data)

            if not is_valid:
                self.logger.error(f"‚ùå –û—à–∏–±–∫–∞ –≤–∞–ª–∏–¥–∞—Ü–∏–∏: {error_msg}")
                processing_time = time.time() - start_time

                return transcription_pb2.TranscriptionResponse(
                    transcript="",
                    success=False,
                    error_message=error_msg,
                    processing_time=processing_time,
                    audio_duration=0.0,
                    stats=transcription_pb2.TranscriptionStats(
                        word_count=0,
                        char_count=0,
                        speed_factor=0.0
                    )
                )

            # –¢—Ä–∞–Ω—Å–∫—Ä–∏–ø—Ü–∏—è —Å –∏—Å–ø–æ–ª—å–∑–æ–≤–∞–Ω–∏–µ–º Borealis –º–æ–¥–µ–ª–∏
            # –°–æ—Ö—Ä–∞–Ω—è–µ–º –∞—É–¥–∏–æ –≤–æ –≤—Ä–µ–º–µ–Ω–Ω—ã–π —Ñ–∞–π–ª
            with tempfile.NamedTemporaryFile(suffix=f".{format_type}", delete=False) as temp_file:
                temp_file.write(audio_data)
                temp_audio_path = temp_file.name

            self.logger.info(f"üíæ –í—Ä–µ–º–µ–Ω–Ω—ã–π —Ñ–∞–π–ª —Å–æ–∑–¥–∞–Ω: {temp_audio_path}")

            # –ó–∞–ø—É—Å–∫–∞–µ–º —Ç—Ä–∞–Ω—Å–∫—Ä–∏–ø—Ü–∏—é —Å –∏—Å–ø–æ–ª—å–∑–æ–≤–∞–Ω–∏–µ–º –∏–Ω—Ç–µ–≥—Ä–∏—Ä–æ–≤–∞–Ω–Ω–æ–π –ª–æ–≥–∏–∫–∏
            transcript = self._transcribe_audio_file(temp_audio_path)

            # –í—ã—á–∏—Å–ª—è–µ–º –¥–ª–∏—Ç–µ–ª—å–Ω–æ—Å—Ç—å –¥–ª—è —Å—Ç–∞—Ç–∏—Å—Ç–∏–∫–∏
            waveform, sr = librosa.load(temp_audio_path, sr=16_000)
            audio_duration = len(waveform) / sr

            # –£–¥–∞–ª—è–µ–º –≤—Ä–µ–º–µ–Ω–Ω—ã–π —Ñ–∞–π–ª
            try:
                Path(temp_audio_path).unlink()
                self.logger.info(f"üóëÔ∏è  –í—Ä–µ–º–µ–Ω–Ω—ã–π —Ñ–∞–π–ª —É–¥–∞–ª–µ–Ω")
            except Exception as e:
                self.logger.warning(f"‚ö†Ô∏è  –ù–µ —É–¥–∞–ª–æ—Å—å —É–¥–∞–ª–∏—Ç—å –≤—Ä–µ–º–µ–Ω–Ω—ã–π —Ñ–∞–π–ª: {e}")

            if transcript is None or transcript == "":
                raise Exception("–¢—Ä–∞–Ω—Å–∫—Ä–∏–ø—Ü–∏—è –≤–µ—Ä–Ω—É–ª–∞ –ø—É—Å—Ç–æ–π —Ä–µ–∑—É–ª—å—Ç–∞—Ç")

            processing_time = time.time() - start_time
            word_count = len(transcript.split())
            char_count = len(transcript)
            speed_factor = audio_duration / processing_time if processing_time > 0 else 0.0

            self.logger.info("=" * 80)
            self.logger.info(f"‚úÖ –°—Ç—Ä–∏–º–∏–Ω–≥–æ–≤–∞—è —Ç—Ä–∞–Ω—Å–∫—Ä–∏–ø—Ü–∏—è —É—Å–ø–µ—à–Ω–æ –∑–∞–≤–µ—Ä—à–µ–Ω–∞!")
            self.logger.info(f"   –û–±—â–µ–µ –≤—Ä–µ–º—è –æ–±—Ä–∞–±–æ—Ç–∫–∏: {processing_time:.2f} —Å–µ–∫")
            self.logger.info(f"   –°–∫–æ—Ä–æ—Å—Ç—å: {speed_factor:.1f}x –æ—Ç —Ä–µ–∞–ª—å–Ω–æ–≥–æ –≤—Ä–µ–º–µ–Ω–∏")
            self.logger.info(f"   –°–ª–æ–≤: {word_count}, –°–∏–º–≤–æ–ª–æ–≤: {char_count}")
            self.logger.info("=" * 80)

            return transcription_pb2.TranscriptionResponse(
                transcript=transcript,
                success=True,
                error_message="",
                processing_time=processing_time,
                audio_duration=audio_duration,
                stats=transcription_pb2.TranscriptionStats(
                    word_count=word_count,
                    char_count=char_count,
                    speed_factor=speed_factor
                )
            )

        except Exception as e:
            error_msg = f"–û—à–∏–±–∫–∞ –ø—Ä–∏ –æ–±—Ä–∞–±–æ—Ç–∫–µ —Å—Ç—Ä–∏–º–∞: {str(e)}"
            self.logger.error(f"‚ùå {error_msg}")
            processing_time = time.time() - start_time

            # –ü—ã—Ç–∞–µ–º—Å—è —É–¥–∞–ª–∏—Ç—å –≤—Ä–µ–º–µ–Ω–Ω—ã–π —Ñ–∞–π–ª –≤ —Å–ª—É—á–∞–µ –æ—à–∏–±–∫–∏
            try:
                if 'temp_audio_path' in locals():
                    Path(temp_audio_path).unlink()
            except:
                pass

            return transcription_pb2.TranscriptionResponse(
                transcript="",
                success=False,
                error_message=error_msg,
                processing_time=processing_time,
                audio_duration=0.0,
                stats=transcription_pb2.TranscriptionStats(
                    word_count=0,
                    char_count=0,
                    speed_factor=0.0
                )
            )

    def _get_transcription_pb2(self):
        """–í–æ–∑–≤—Ä–∞—â–∞–µ—Ç –º–æ–¥—É–ª—å protobuf –¥–ª—è –≤–µ—Ä—Å–∏–∏ v1"""
        return transcription_pb2

    def get_version(self) -> str:
        """–í–æ–∑–≤—Ä–∞—â–∞–µ—Ç –≤–µ—Ä—Å–∏—é —Å–µ—Ä–≤–∏—Å–∞"""
        return "borealis"
